chainindex/helpers.go

```diff
@@ -0,0 +1,108 @@
+package chainindex
+
+import (
+	"context"
+	"database/sql"
+	"os"
+
+	"golang.org/x/xerrors"
+
+	"github.com/filecoin-project/lotus/api"
+	"github.com/filecoin-project/lotus/chain/types"
+)
+
+func PopulateFromSnapshot(ctx context.Context, path string, cs ChainStore) error {
+	// if a database already exists, we try to delete it and create a new one
+	if _, err := os.Stat(path); err == nil {
+		if err = os.Remove(path); err != nil {
+			return xerrors.Errorf("chainindex already exists at %s and can't be deleted", path)
+		}
+	}
+
+	si, err := NewSqliteIndexer(path, cs, 0)
+	if err != nil {
+		return xerrors.Errorf("failed to create sqlite indexer: %w", err)
+	}
+	defer func() {
+		_ = si.Close()
+	}()
+
+	err = withTx(ctx, si.db, func(tx *sql.Tx) error {
+		curTs := cs.GetHeaviestTipSet()
+		startHeight := curTs.Height()
+
+		for curTs != nil {
+			if err := si.indexTipset(ctx, tx, curTs); err != nil {
+				log.Infof("stopping import after %d tipsets", startHeight-curTs.Height())
+				break
+			}
+
+			curTs, err = cs.GetTipSetFromKey(ctx, curTs.Parents())
+			if err != nil {
+				log.Infof("stopping import after %d tipsets", startHeight-curTs.Height())
+				break
+			}
+		}
+
+		return nil
+	})
+	if err != nil {
+		return xerrors.Errorf("failed populate from snapshot: %w", err)
+	}
+
+	return nil
+}
+
+func WaitForMpoolUpdates(ctx context.Context, ch <-chan api.MpoolUpdate, indexer Indexer) {
+	for {
```

Comment:
```suggestion
	for ctx.Err() == nil {
```

don't need to drain the channel if context is cancelled do we?